alf/experience_replayers/replay_buffer.py

```diff
@@ -132,48 +120,43 @@ def get_batch(self, batch_size, batch_length):
 
             batch_size_per_env = batch_size // self._num_envs
             remaining = batch_size % self._num_envs
             if batch_size_per_env > 0:
-                env_ids = tf.tile(
-                    tf.range(self._num_envs, dtype=tf.int64),
-                    [batch_size_per_env])
+                env_ids = torch.arange(
+                    self._num_envs).repeat(batch_size_per_env)
             else:
-                env_ids = tf.zeros((0, ), tf.int64)
+                env_ids = torch.zeros(0, dtype=torch.int64)
             if remaining > 0:
-                eids = tf.range(self._num_envs, dtype=tf.int64)
-                eids = tf.random.shuffle(eids)[:remaining]
-                env_ids = tf.concat([env_ids, eids], axis=0)
+                eids = torch.randperm(self._num_envs)[:remaining]
+                env_ids = torch.cat([env_ids, eids], dim=0)
 
-            r = tf.random.uniform(tf.shape(env_ids))
+            r = torch.rand(*env_ids.shape)
             num_positions = self._current_size - batch_length + 1
-            num_positions = tf.gather(num_positions, env_ids)
-            pos = tf.cast(r * tf.cast(num_positions, tf.float32), tf.int64)
-            pos += tf.gather(self._current_pos - self._current_size, env_ids)
-            pos = tf.reshape(pos, [-1, 1])  # [B, 1]
-            pos = pos + tf.expand_dims(
-                tf.range(batch_length, dtype=tf.int64), axis=0)  # [B, T]
+            num_positions = num_positions[env_ids]
+            pos = (r * num_positions).to(torch.int64)
+            pos += (self._current_pos - self._current_size)[env_ids]
+            pos = pos.reshape(-1, 1)  # [B, 1]
+            pos = pos + torch.arange(batch_length).unsqueeze(0)  # [B, T]
             pos = pos % self._max_length
-            pos = tf.expand_dims(pos, -1)  # [B, T, 1]
-            env_ids = tf.reshape(env_ids, [-1, 1])  # [B, 1]
-            env_ids = tf.tile(env_ids, [1, batch_length])  # [B, T]
-            env_ids = tf.reshape(env_ids, [-1, batch_length, 1])  # [B, T, 1]
-            indices = tf.concat([env_ids, pos], axis=-1)  # [B, T, 2]
-            return tf.nest.map_structure(
-                lambda buffer: tf.gather_nd(buffer, indices), self._buffer)
+            env_ids = env_ids.reshape(-1, 1).repeat(1, batch_length)  # [B, T]
+            indices = env_ids * self._max_length + pos  # [B, T]
+            indices = indices.reshape(-1)  # [B*T]
+            return alf.nest.map_structure(
+                lambda buffer: buffer[indices].reshape(
+                    batch_size, batch_length, *buffer.shape[1:]),
+                self._flattened_buffer)
 
     def clear(self):
-        with tf.device(self._device):
-            self._current_size.assign(tf.zeros_like(self._current_size))
-            self._current_pos.assign(tf.zeros_like(self._current_pos))
+        self._current_size.assign(torch.zeros_like(self._current_size))
+        self._current_pos.assign(tf.zeros_like(self._current_pos))
```

Comment:
tf -> torch; missing test cases for `clear`

Reply:
Fixed